### PR TITLE
refactor: use RxJs for WaitForNetworkIdle

### PR DESCRIPTION
**Pull request by Lightning00Blade**

### CODE REVIEW

File: packages/puppeteer-core/src/api/Page.ts [hunk 1]
```diff
@@ -1696,62 +1700,33 @@ export abstract class Page
...
+        fromEvent(
+          networkManager,
+          NetworkManagerEmittedEvents.Response as unknown as string
+        ),
+        fromEvent(
+          networkManager,
+          NetworkManagerEmittedEvents.RequestFailed as unknown as string
+        )
+      ).pipe(
+        startWith(null),
```
Comment (jrandolf) [inline packages/puppeteer-core/src/api/Page.ts:1721]: Why `startWith`?

File: packages/puppeteer-core/src/api/Page.ts [hunk 2]
```diff
@@ -1696,62 +1700,33 @@ export abstract class Page
...
+        filter(() => {
+          return networkManager.inFlightRequestsCount() === 0;
+        }),
+        switchMap(v => {
```
Comment (jrandolf) [inline packages/puppeteer-core/src/api/Page.ts:1725]: Use `mergeMap`.

File: packages/puppeteer-core/src/common/util.ts [hunk 3]
```diff
@@ -708,3 +710,16 @@ export class Mutex {
     resolve();
   }
 }
+
+/**
+ * @internal
+ */
+export function timeout(ms: number): Observable<never> {
+  return ms === 0
```
Comment (jrandolf) [inline packages/puppeteer-core/src/common/util.ts:718]: How about `Infinity` rather than `0`?

**Review by jrandolf-google — APPROVED** [review]

**Review by jrandolf-google — APPROVED** [review]
lgtm % comments

File: packages/puppeteer-core/src/api/Page.ts (hunk 1 above)
Comment (Lightning00Blade) [reply packages/puppeteer-core/src/api/Page.ts:1721]: An event may never come we need to trigger the observer 1 time.

File: packages/puppeteer-core/src/common/util.ts (hunk 3 above)
Comment (Lightning00Blade) [reply packages/puppeteer-core/src/common/util.ts:718]: That's incorrect as the Common Puppeteer representation of `No Timeout` is `timeout=0`. 
Also if you use `SetTimeout(()=>{}, Infinity)` Node may complain.
```
TimeoutOverflowWarning: Infinity does not fit into a 32-bit signed integer.
```

File: packages/puppeteer-core/src/api/Page.ts (hunk 2 above)
Comment (Lightning00Blade) [reply packages/puppeteer-core/src/api/Page.ts:1725]: That does not work. We need `switchMap` as we want to create a new Observable that will be cancelled if its we rerun `swichMap` and produce a new instance.

**Pull request merged**